ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Protocol/Crypto/HotKey.hs

```diff
@@ -78,37 +80,41 @@ toPeriod HotKey { hkStart = Absolute.KESPeriod lo, hkEvolution } =
 --
 -- When the given KES period is outside the bounds of the 'HotKey', we return
 -- 'Nothing'.
+--
+-- The old 'HotKey' (and an intermediary ones produced while evolving) is
+-- securely erased from memory.
 evolve ::
      forall c m. (IOLike m, Crypto c)
   => Absolute.KESPeriod
   -> HotKey c
   -> m (Maybe (HotKey c))
-evolve targetPeriod hk = do
-    let mNewHotKey = do
-           targetEvolution <- toEvolution hk targetPeriod
-           key' <- go targetEvolution (hkEvolution hk) (hkKey hk)
-           return HotKey {
-               hkStart     = hkStart hk
-             , hkEnd       = hkEnd   hk
-             , hkEvolution = targetEvolution
-             , hkKey       = key'
-             }
-    -- Any stateful code (for instance, running finalizers to clear the
-    -- memory associated with the old key) would happen here or in (a
-    -- monadic) 'go'.
-    return mNewHotKey
+evolve targetPeriod hk
+  | Just targetEvolution <- toEvolution hk targetPeriod
+  = go targetEvolution (hkEvolution hk) (hkKey hk) >>= \case
+      Nothing   -> return Nothing
+      Just key' -> return $ Just HotKey {
+          hkStart     = hkStart hk
+        , hkEnd       = hkEnd   hk
+        , hkEvolution = targetEvolution
+        , hkKey       = key'
+        }
+  | otherwise
+  = return Nothing
   where
```

Comment:
Mask in https://github.com/input-output-hk/ouroboros-network/blob/4cedb802007f44d57a5b59e7eb89d1172a2fd2ca/ouroboros-consensus/src/Ouroboros/Consensus/Node/BlockProduction.hs#L87 document this.